### ISSUE TITLE
Batch error on REST api with Neo4j 2.1.1

### ISSUE THREAD

**Issue by benitoDeLaCasita** [neo4j/neo4j#2517]
On a Debian wheezy instance I've just migrated my standalone Neo4j server from version 2.0.2 to 2.1.1 and I'm experiencing problems with batch queries on REST api :

On a empty database I tried the following test :

* First I created a node :

```
curl -H Accept:application/json -H Content-Type:application/json -X POST -d '{"name":"Steven"}' http://localhost:7474/db/data/node
```

the node was created successfully :

```json
{
  "extensions": {},
  "paged_traverse": "http://localhost:7474/db/data/node/0/paged/traverse/{returnType}{?pageSize,leaseTime}",
  "labels": "http://localhost:7474/db/data/node/0/labels",
  "outgoing_relationships": "http://localhost:7474/db/data/node/0/relationships/out",
  "traverse": "http://localhost:7474/db/data/node/0/traverse/{returnType}",
  "all_typed_relationships": "http://localhost:7474/db/data/node/0/relationships/all/{-list|&|types}",
  "property": "http://localhost:7474/db/data/node/0/properties/{key}",
  "all_relationships": "http://localhost:7474/db/data/node/0/relationships/all",
  "self": "http://localhost:7474/db/data/node/0",
  "outgoing_typed_relationships": "http://localhost:7474/db/data/node/0/relationships/out/{-list|&|types}",
  "properties": "http://localhost:7474/db/data/node/0/properties",
  "incoming_relationships": "http://localhost:7474/db/data/node/0/relationships/in",
  "incoming_typed_relationships": "http://localhost:7474/db/data/node/0/relationships/in/{-list|&|types}",
  "create_relationship": "http://localhost:7474/db/data/node/0/relationships",
  "data": {
    "name": "Steven"
  }
}

```

* then I tried to get the new node :

```
curl -H Accept:application/json -H Content-Type:application/json -X GET http://localhost:7474/db/data/node/0
```

which gave me the expected result :

```json
{
  "extensions": {},
  "paged_traverse": "http://localhost:7474/db/data/node/0/paged/traverse/{returnType}{?pageSize,leaseTime}",
  "labels": "http://localhost:7474/db/data/node/0/labels",
  "outgoing_relationships": "http://localhost:7474/db/data/node/0/relationships/out",
  "traverse": "http://localhost:7474/db/data/node/0/traverse/{returnType}",
  "all_typed_relationships": "http://localhost:7474/db/data/node/0/relationships/all/{-list|&|types}",
  "property": "http://localhost:7474/db/data/node/0/properties/{key}",
  "all_relationships": "http://localhost:7474/db/data/node/0/relationships/all",
  "self": "http://localhost:7474/db/data/node/0",
  "outgoing_typed_relationships": "http://localhost:7474/db/data/node/0/relationships/out/{-list|&|types}",
  "properties": "http://localhost:7474/db/data/node/0/properties",
  "incoming_relationships": "http://localhost:7474/db/data/node/0/relationships/in",
  "incoming_typed_relationships": "http://localhost:7474/db/data/node/0/relationships/in/{-list|&|types}",
  "create_relationship": "http://localhost:7474/db/data/node/0/relationships",
  "data": {
    "name": "Steven"
  }
}
```

* but the same request in a batch gave me the following error

```
curl -H Accept:application/json -H Content-Type:application/json -X POST -d '[{"method":"GET","to":"/node/0","id":0}]' http://localhost:7474/db/data/batch
```

```json
{
  "exception": "NullPointerException",
  "fullname": "java.lang.NullPointerException",
  "stacktrace": [
    "org.neo4j.server.web.Jetty9WebServer.invokeDirectly(Jetty9WebServer.java:339)",
    "org.neo4j.server.rest.batch.NonStreamingBatchOperations.invoke(NonStreamingBatchOperations.java:55)",
    "org.neo4j.server.rest.batch.BatchOperations.performRequest(BatchOperations.java:201)",
    "org.neo4j.server.rest.batch.BatchOperations.parseAndPerform(BatchOperations.java:171)",
    "org.neo4j.server.rest.batch.NonStreamingBatchOperations.performBatchJobs(NonStreamingBatchOperations.java:48)",
    "org.neo4j.server.rest.web.BatchOperationService.batchProcess(BatchOperationService.java:126)",
    "org.neo4j.server.rest.web.BatchOperationService.performBatchOperations(BatchOperationService.java:76)",
    "java.lang.reflect.Method.invoke(Method.java:606)",
    "org.neo4j.server.rest.transactional.TransactionalRequestDispatcher.dispatch(TransactionalRequestDispatcher.java:139)",
    "java.lang.Thread.run(Thread.java:744)"
  ]
}
```

The only log I saw was the http ones :

```
127.0.0.1 - - [03/Jun/2014:14:56:17 +0200] "POST /db/data/node HTTP/1.1" 201 1160 "-" "curl/7.26.0"
127.0.0.1 - - [03/Jun/2014:14:56:33 +0200] "GET /db/data/node/0 HTTP/1.1" 200 1160 "-" "curl/7.26.0"
127.0.0.1 - - [03/Jun/2014:14:56:47 +0200] "POST /db/data/batch HTTP/1.1" 500 1000 "-" "curl/7.26.0"
```

I run on java-7-openjdk-amd64

Any help would be appreciated.

Thank you.

Edit: The exact same request in streaming mode (i.e. with header X-Stream set to true) has a different behavior : it gives the expected data but with a status 500.

```
curl -H X-Stream:true -H Accept:application/json -H Content-Type:application/json -X POST -d '[{"method":"GET","to":"/node/0","id":0}]' http://localhost:7474/db/data/batch
```
```json
[
  {
    "id":0,
    "from":"/node/0",
    "body":
      {
        "extensions":{},
        "paged_traverse":"http://localhost:7474/db/data/node/0/paged/traverse/{returnType}{?pageSize,leaseTime}",
        "labels":"http://localhost:7474/db/data/node/0/labels",
        "outgoing_relationships":"http://localhost:7474/db/data/node/0/relationships/out",
        "traverse":"http://localhost:7474/db/data/node/0/traverse/{returnType}",
        "all_typed_relationships":"http://localhost:7474/db/data/node/0/relationships/all/{-list|&|types}",
        "property":"http://localhost:7474/db/data/node/0/properties/{key}",
        "all_relationships":"http://localhost:7474/db/data/node/0/relationships/all",
        "self":"http://localhost:7474/db/data/node/0",
        "properties":"http://localhost:7474/db/data/node/0/properties",
        "outgoing_typed_relationships":"http://localhost:7474/db/data/node/0/relationships/out/{-list|&|types}",
        "incoming_relationships":"http://localhost:7474/db/data/node/0/relationships/in",
        "incoming_typed_relationships":"http://localhost:7474/db/data/node/0/relationships/in/{-list|&|types}",
        "create_relationship":"http://localhost:7474/db/data/node/0/relationships",
        "data":
          {
            "name":"Steven"
          }
      },
    "status":500
  }
]
```

**Comment by glowskir** [comment]
Bump, still seen in 2.1.7, at some point even simplest batch operation fails
```json
[ {
  "to" : "cypher",
  "body" : {
    "query" : "MATCH (n) RETURN n limit 1",
    "params" : {
      "uuid" : "fdfb14c0-cd91-11e4-a58a-b33c4994e7d9"
    }
  },
  "method" : "POST"
}]

{
    "exception": "NullPointerException",
    "fullname": "java.lang.NullPointerException",
    "stacktrace": [
        "org.neo4j.server.web.Jetty9WebServer.invokeDirectly(Jetty9WebServer.java:340)",
        "org.neo4j.server.rest.batch.NonStreamingBatchOperations.invoke(NonStreamingBatchOperations.java:56)",
        "org.neo4j.server.rest.batch.BatchOperations.performRequest(BatchOperations.java:205)",
        "org.neo4j.server.rest.batch.BatchOperations.parseAndPerform(BatchOperations.java:173)",
        "org.neo4j.server.rest.batch.NonStreamingBatchOperations.performBatchJobs(NonStreamingBatchOperations.java:49)",
        "org.neo4j.server.rest.web.BatchOperationService.batchProcess(BatchOperationService.java:128)",
        "org.neo4j.server.rest.web.BatchOperationService.performBatchOperations(BatchOperationService.java:77)",
        "java.lang.reflect.Method.invoke(Method.java:606)",
        "org.neo4j.server.rest.transactional.TransactionalRequestDispatcher.dispatch(TransactionalRequestDispatcher.java:139)",
        "ch.qos.logback.access.servlet.TeeFilter.doFilter(TeeFilter.java:55)",
        "java.lang.Thread.run(Thread.java:745)"
    ]
}

**Comment by glowskir** [comment]
apparently this was caused by `org.neo4j.server.http.log.enabled=true`. **facepalm**

**Comment by rishabhjain08** [comment]
Do you have any idea why this config can cause this exception. I have been battling with it for a long time.

**Comment by glowskir** [comment]
Nope, trial and error, haven't got time to figure it out exactly.

**Comment by charlieanstey** [comment]
I've been having issues with batch queries too, throwing a Neo.DatabaseError.General.UnknownFailure wrapped in an NPE. Pure luck I found this issue. And yes, disabling HTTP logging removes the error?!? Crazy one.

**Comment by RaphHaddad** [comment]
Any status update on this one yet?

**Comment by destronije** [comment]
Someone should really take a look at this.

I've been fighting with this issue for 2 days, and found this issue by pure luck like @charlieanstey said.

**Comment by jexp** [comment]
If possible you should not try and use the /batch endpoint anymore. But rather /db/data/transaction with multiple statements.

And yes it is related to http logging, as we "fake" an internal jetty request and some of the information that is needed for logging is missing there.

This particular issue should have been fixed in 2.2 with this PR

https://github.com/neo4j/neo4j/pull/5049
